lldb/tools/lldb-server/Plugins/AMDGPU/ProcessAMDGPU.cpp

```diff
@@ -437,8 +486,195 @@ bool ProcessAMDGPU::handleDebugEvent(amd_dbgapi_event_id_t eventId,
+                                      &wave_info.queue_id))
+    return err;
+
+  if (llvm::Error err = QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_AGENT,
+                                      &wave_info.agent_id))
+    return err;
+
+  if (llvm::Error err = QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_PROCESS,
+                                      &wave_info.process_id))
+    return err;
+
+  if (llvm::Error err =
+          QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_ARCHITECTURE,
+                        &wave_info.architecture_id))
+    return err;
+
+  if (llvm::Error err =
+          QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_WORKGROUP_COORD,
+                        &wave_info.workgroup_coord))
+    return err;
+
+  if (llvm::Error err =
+          QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_WAVE_NUMBER_IN_WORKGROUP,
+                        &wave_info.index_in_workgroup))
+    return err;
+
+  if (llvm::Error err = QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_LANE_COUNT,
+                                      &wave_info.num_lanes_supported))
+    return err;
+
+  // Some information can only be queried if the wave is stopped.
+  if (wave_info.state == AMD_DBGAPI_WAVE_STATE_STOP) {
+    if (llvm::Error err = QueryWaveInfo(
+            wave_id, AMD_DBGAPI_WAVE_INFO_STOP_REASON, &wave_info.stop_reason))
+      return err;
+
+    if (llvm::Error err =
+            QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_PC, &wave_info.pc))
+      return err;
+
+    if (llvm::Error err = QueryWaveInfo(wave_id, AMD_DBGAPI_WAVE_INFO_EXEC_MASK,
+                                        &wave_info.exec_mask))
+      return err;
+  }
+
+  return wave_info;
+}
+
+llvm::Expected<DbgApiClientMemoryPtr<amd_dbgapi_wave_id_t>>
+ProcessAMDGPU::GetWaveList(size_t *count, amd_dbgapi_changed_t *changed) {
+  // Re-enable wave creation on exit from this function.
+  auto ResetWaveCreation = llvm::make_scope_exit([this] {
+    if (auto error = RunAmdDbgApiCommand([this] {
+          return amd_dbgapi_process_set_wave_creation(
+              GetDbgApiProcessID(), AMD_DBGAPI_WAVE_CREATION_NORMAL);
+        }))
+      LLDB_LOG_ERROR(GetLog(GDBRLog::Plugin), std::move(error),
+                     "Error: Failed to enable wave creation: {0}");
+  });
+
+  // Stop creating new waves get an accurate count.
+  if (llvm::Error error = RunAmdDbgApiCommand([this] {
+        return amd_dbgapi_process_set_wave_creation(
+            GetDbgApiProcessID(), AMD_DBGAPI_WAVE_CREATION_STOP);
+      }))
+    return error;
+
+  // Get the list of waves
+  amd_dbgapi_wave_id_t *wave_list = nullptr;
+  if (auto error = RunAmdDbgApiCommand([&] {
+        return amd_dbgapi_process_wave_list(GetDbgApiProcessID(), count,
+                                            &wave_list, changed);
+      }))
+    return error;
+
+  return DbgApiClientMemoryPtr<amd_dbgapi_wave_id_t>(wave_list);
+}
+
+WaveIdList ProcessAMDGPU::UpdateWaves() {
```

Comment:
It wasn't clear to me that this was returning only newly added waves. Maybe a clearer name would name this more obvious?
```
WaveIdList ProcessAMDGPU::UpdateWavesAndReturnNewWaves();
```
or 
```
void ProcessAMDGPU::UpdateWaves(WaveIdList &new_waves);
```